app/src/ingest_edd_web.py

```diff
@@ -32,18 +32,47 @@ def _ingest_edd_web(
         urls_processed.add(item["url"])
 
         content = item.get("main_content", item.get("main_primary"))
-        assert content, f"Item {name} has no main_content or main_primary"
+        if not content:
+            logger.warning("Item %s has no main_content or main_primary", name)
+            continue
 
-        document = Document(name=name, content=content, **doc_attribs)
+        document = Document(name=name, content=content, source=item["url"], **doc_attribs)
         db_session.add(document)
 
-        embedding_model = app_config.sentence_transformer
-        tokens = len(embedding_model.tokenizer.tokenize(content))
-        mpnet_embedding = embedding_model.encode(content, show_progress_bar=False)
-        chunk = Chunk(
-            document=document, content=content, tokens=tokens, mpnet_embedding=mpnet_embedding
+        chunks = _chunk_page(content)
+        for chunk in chunks:
+            chunk.document = document
+        db_session.add_all(chunks)
+        logger.info("Split %s into %d chunks", name, len(chunks))
+
+
+def _chunk_page(content: str) -> list[Chunk]:
+    embedding_model = app_config.sentence_transformer
+
+    # Split content by double newlines, then gather into the largest chunks
+    # that tokenize to less than the max_seq_length
+    content_split_by_double_newlines = content.split("\n\n")
+    subsections = [content_split_by_double_newlines[0]]
+    for subsection in content_split_by_double_newlines[1:]:
+        tokens_with_new_subsection = len(
+            embedding_model.tokenizer.tokenize(subsections[-1] + subsection)
```

Comment:
Much cleaner code. For speed, I tried to reduce the number of calls to the embedding model by estimating the number of target chunks but the advantages is not yet clear.
https://github.com/navapbc/labs-decision-support-tool/blob/acadee85a8beee68d5aafbbab1bb90d1955f6c63/app/src/ingest_bem_pdfs.py#L201-L206

Reply:
Oh right, thanks for reminding me about that code -- I wonder how expensive `embedding_model.tokenizer.tokenize` is

Reply:
If you use or adapt that code, start with a lower multiplication factor than `1.5`, like `1.2`.